### ISSUE TITLE
Vizdoom navigation problem

### ISSUE THREAD

**Issue by pengzhi1998** [mwydmuch/ViZDoom#445]
Hi, I'm using VizDoom to train my agent to learn to navigate. But there are several questions: 
1. in my scenario, the maze is discrete, which means I regard the vizdoom as a 3D grid world. However, I don't know how to randomly initialize my agent in those grids. 
2. By the way, I want the agent to move in three ways: move forward for 1 grid (a specific length), turn left (90 degrees), and turn right (90 degrees as well). But I only find the button to move in a very small step or turn a small angle. 
Could you please provide me with some hints to do those things? Thanks a lot!

**Comment by pengzhi1998** [comment]
Hi, I'm using VizDoom to train my agent to learn to navigate. But there are several questions: 1. in my scenario, the maze is discreted, which means I regard the vizdoom as a 3D grid world. However, I don't know how to randomly initialize my agent in those grids. 2. By the way, I want the agent to move in three ways: move forward for 1 grid (a specific length), turn left (90 degrees) and turn right (90 degrees as well). But I only find the button to move in a very small step or turn a small angle. Could you please provide some hints to do those things? Thanks a lot!

**Comment by Miffyli** [comment]
1) Take a look at [MazeExplorer](https://github.com/microsoft/MazeExplorer) to generate random levels with ViZDoom
2) You can use `Button.TURN_LEFT_RIGHT_DELTA` to turn specific number of degrees (the value is degrees-per-step, so setting it to 90 and proceeding by one step will turn 90 degrees). You need to manually measure how much you need to forward and backward to step by one "grid-block".

**Comment by pengzhi1998** [comment]
Thank you so much!

**Comment by mwydmuch** [comment]
Hi @pengzhi1998, as @Miffyli said, you can use `Button.TURN_LEFT_RIGHT_DELTA` and `Button.MOVE_FORWARD_BACKWARD_DELTA` to simulate discrete movement. Values you passed as an action can be any floating point number, while for normal buttons value other than 0 is interpreted as pushed, for delta buttons value actually determine an angle/distance. So for `Button.TURN_LEFT_RIGHT_DELTA` you can pass value -90/90 to turn by 90 degrees left/right in one step. This example presents this functionality: https://github.com/mwydmuch/ViZDoom/blob/master/examples/python/delta_buttons.py

**Comment by pengzhi1998** [comment]
@mwydmuch Thank you!!

**Comment by pengzhi1998** [comment]
By the way, I checked the issues and found some researchers facing problems (such as memory leak) when using VizDoom to train A3C models with PyTorch. Is it still a big problem for Pytorch on multi-processing programs in VizDoom nowadays?

**Comment by Miffyli** [comment]
ViZDoom should not be leaking memory when used correctly. I have not had issues using it with either PyTorch or Tensorflow. Mind you, I have not tried using the pytorch.multiprocessing much.

If you repeatedly call `game.close()` and `game.init()`, that is one possible source of memory leakage and I would advice against it.

**Comment by mwydmuch** [comment]
@pengzhi1998 personally I've never noticed a significant increase in memory usage by ViZDoom (and it uses just a few mega) even after many hours of running the same instance that was playing the same scenario over and over again. 

But as @Miffyli, I also discourage a usage of `game.init()` and `game.close()`. It's better to use one or a few running instances as long as possible and restarting the scenario with `game.new_episode()`. This will be also much faster since `game.new_episode()` takes a few ms while closing and starting the whole engine again takes almost a second. 

Also, to correctly handle restart called from ACS script you need to check if `game.is_episode_finished()` and if true then call `game.new_episode()`.

**Comment by mwydmuch** [comment]
Oh, and a good way to finish an episode from ACS script is to kill a player with `https://zdoom.org/wiki/DamageActor` 😄

**Comment by mwydmuch** [comment]
Sorry, I forgot, actually there is a better way, `Exit_Normal(0)` function that will finish the episode.

**Comment by pengzhi1998** [comment]
Thank you so much! It's so nice of you!!
Nevertheless, I need to start a new episode after the agent takes a specific number (for example, 30) of steps. However, the step counter is set in the python script, but not in the ACS. Specifically speaking, what I'd like to do is starting a new episode when the python program takes 30 cycles of a loop (the agent takes 30 actions). The biggest problem for me now is that I'm not sure how to trigger a new episode based on the number of steps but not events in the Doom games (for example, kill a monster or reach a goal). Do you have any suggestions?

**Comment by pengzhi1998** [comment]
I found a possible solution [here](https://github.com/devendrachaplot/DeepRL-Grounding/blob/master/env.py#L51). The researcher just uses `game.new_episode()` to start a new episode without checking `game.is_episode_finished()` or setting `restart` inside their ACS. They just use the step counter in their python program and start a new episode which will base on the counter. Not sure whether it's a good way. But I'll take a try.

**Comment by mwydmuch** [comment]
@pengzhi1998 yeah you can use `game.new_episode()` at any point to restart, sorry, I thought that it's quite obvious. 

When it comes to memory leak, I ran this code that you posted (with my_way_home.wad) and I didn't see any memory leak, amount of ram used by ViZDoom was the same for the whole time. Can you provide more details about the issue? How fast is the increese in memory usage, what OS and Python version you are using, can you provide this wad file?

**Comment by pengzhi1998** [comment]
Thank you! 
I'm sorry I thought the new episode trigger had to be set inside the ACS (for example, the game takes a new episode only when the player kills a monster or approaches a goal point). So I thought I had to set something manually in ACS or use the `is_episode_finished()` function. Thank you for telling me that!
About the issue of memory problem, I found that in 1000 episodes, the program increasingly takes around 1 gigabyte which is wired. I tested it on Ubuntu 18.04.2 LTS with python 3.6.10. The ACS file is shown below:
`#include "zcommon.acs"

#define TARGET_ID_START 1000 
#define GOAL_TID 999

global int 0:reward;
global int 1:goal_x;
global int 2:goal_y;
global int 3:goal_z;

int TARGET_ID_END = TARGET_ID_START;
int SPAWN_LOC_ID = 0;
int GOAL_LOC_ID = 0;

function int fdistance (int tid1, int tid2)
{
	int len;
	int y = getactory(tid1) - getactory(tid2);
	int x = getactorx(tid1) - getactorx(tid2);
	int z = getactorz(tid1) - getactorz(tid2);

	int ang = vectorangle(x,y);
	if(((ang+0.125)%0.5) > 0.25) len = fixeddiv(y, sin(ang));
	else len = fixeddiv(x, cos(ang));

	ang = vectorangle(len, z);
	if(((ang+0.125)%0.5) > 0.25) len = fixeddiv(z, sin(ang));
	else len = fixeddiv(len, cos(ang));

	return len;
}

script 1 ENTER
{
	TARGET_ID_END = TARGET_ID_START;
	while(IsTIDUsed(TARGET_ID_END + 1))
	{
		TARGET_ID_END += 1;
	}

	// Spawn actor
	SPAWN_LOC_ID = random(TARGET_ID_START, TARGET_ID_END);
	SetActorPosition(0, GetActorX(SPAWN_LOC_ID), GetActorY(SPAWN_LOC_ID), 0.0, 0);
	SetActorAngle(0, 1.0);
	SetActorVelocity(0, 0, 0, 0, FALSE, FALSE);

	// Spawn goals
	if(!IsTIDUsed(GOAL_TID)) {
		GOAL_LOC_ID = SPAWN_LOC_ID;
		until(GOAL_LOC_ID!=SPAWN_LOC_ID) GOAL_LOC_ID = random(TARGET_ID_START, TARGET_ID_END);
		until(SpawnSpot("TallRedColumn", GOAL_LOC_ID, GOAL_TID));
		goal_x = GetActorX(GOAL_LOC_ID);
		goal_y = GetActorY(GOAL_LOC_ID);
		goal_z = GetActorZ(GOAL_LOC_ID);
	}

	// Wait
	until(fdistance(0, GOAL_TID) < 64.0) 
	{
		Delay(1);
	}
	reward += 10.0;

	// restart;
}`

**Comment by Miffyli** [comment]
1) I ran the scenario you provided with a random agent, and did not observe any memory leakage (in fact, memory use was a very fixed number). Ubuntu 18.04, Python 3.6, ViZDoom 1.1.8 (current `master`).

2) Yes, there are some limitations for the speed, and the fact there is acceleration involved makes things more difficult, too. You could measure how long it takes to move for one block when pressing `Button.FORWARD`, and then use that tick amount together with `make_action` to move correct amount. This could be possible with fancy ACS scripts that teleport player around, but I do not know how this could be done.

**Comment by mwydmuch** [comment]
It should be possible to build scenario with discrete movement using https://zdoom.org/wiki/GetPlayerInput in a while loop for checking inputs and moving the player with https://zdoom.org/wiki/SetActorPosition, but probably some collision checking will be necessary. It can be done by checking player's sector and it's lines or have some dummy actor teleport it to the player's destination and check https://zdoom.org/wiki/CheckSight. That would be my approach to it.

**Comment by pengzhi1998** [comment]
Here is a very simple version of the code snippet:
```ACS
while(True)
	{
	    int buttons = GetPlayerInput(-1, INPUT_BUTTONS);
	    if(buttons == BT_FORWARD)
	    {
	        print(s:"Hello!");
	        SetActorPosition(0, GetActorX(SPAWN_LOC_ID) + 100, GetActorY(SPAWN_LOC_ID), 0.0, 0);
	    }
		Delay(1);
	}
```
But the player cannot be set into the new place (x position added with 100). 
If the problem could be solved, I'm still wondering about another situation: when the agent chooses the action to move forward, even though its position could be teleported by `SetActorPosition`, the `move_forward` action will be executed anyhow, which might lead to the agent not in the exact wanted position. How can I make sure that when the forward action is chosen, the action itself will not be executed, but just execute the `SetActorPosition` function?
The full ACS is shown here:
```ACS
#include "zcommon.acs"

#define TARGET_ID_START 1000 
#define GOAL_TID 999

global int 0:reward;
global int 1:goal_x;
global int 2:goal_y;
global int 3:goal_z;

int TARGET_ID_END = TARGET_ID_START;
int SPAWN_LOC_ID = 0;
int GOAL_LOC_ID = 0;

script 1 ENTER
{

	TARGET_ID_END = TARGET_ID_START;
	while(IsTIDUsed(TARGET_ID_END + 1))
	{
		TARGET_ID_END += 1;
	}

    int a = random(0, 100);
    print(d:a);
	// Spawn actor
	SPAWN_LOC_ID = random(TARGET_ID_START, TARGET_ID_END);
	SetActorPosition(0, GetActorX(SPAWN_LOC_ID), GetActorY(SPAWN_LOC_ID), 0.0, 0);
	if (a < 25) {
	    SetActorAngle(0, 0.);
	} else if (a < 50) {
	    SetActorAngle(0, 0.25);
	} else if (a < 75) {
	    SetActorAngle(0, 0.5);
	} else {
	    SetActorAngle(0, 0.75);
	}
	SetActorVelocity(0, 0, 0, 0, FALSE, FALSE);

	// Wait
	while(True)
	{
	    int buttons = GetPlayerInput(-1, INPUT_BUTTONS);
	    if(buttons == BT_FORWARD)
	    {
	        print(s:"Hello!");
	        SetActorPosition(0, GetActorX(SPAWN_LOC_ID) + 100, GetActorY(SPAWN_LOC_ID), 0.0, 0);
	    }
		Delay(1);
	}
	reward += 10.0;
}```

**Comment by pengzhi1998** [comment]
Sorry to bother you again. 
Actually I have solved the former problems. Thank you so much for your great help! 
I have another question about the position. When running one python script, the position is shown all in positive numbers. However, in another script, the position is shown sometimes in negative numbers. It seems like the coordinate system has been changed when running different python script codes. However, I set the variables in the same way:
`game.add_available_game_variable(GameVariable.POSITION_X)`
And I just used the same WAD file. Do you know how to deal with this problem?

**Comment by Miffyli** [comment]
There's nothing stopping the coordinates from being negative. It all depends on how the map was built and where the players spawn. Origin can be in one of the corners of the map, or maybe in the center. You can process this much like positive numbers.

**Comment by pengzhi1998** [comment]
Thank you for your reply!
However, the wired thing is that I'm using the same WAD file, which means the place the player is spawned is just the same. Because I need to deal with the global position of the agent, the inconsistency is a problem. So is there a way to define the origin? Or just figure out where the origin is?

**Comment by mwydmuch** [comment]
This sound strange, the position should be consistent. Are you sure the player is not spawned/teleported to the different locations each time you start your scenario? I can look into this if you provide the scripts and WAD file to replicate the problem.

**Comment by pengzhi1998** [comment]
Hi, thank you both so much!!
I cannot solve those problems without your help. Thanks again!
Actually, I'm trying to deal with a localization problem with VizDoom in my undergraduate thesis. I wrote two scripts: [example.py](https://github.com/pengzhi1998/Localization-VizDoom/blob/master/Create_VizDoom_Env/example.py) and [static_goal.acs](https://github.com/pengzhi1998/Localization-VizDoom/blob/master/Create_VizDoom_Env/content/static_goal.acs) which have solved the mentioned problems above. And they are borrowed from your project as well as NavDoom. Hopefully, if other programmers face similar problems, the scripts might help.
